<a href="https://colab.research.google.com/github/accarter/DS-Unit-2-Linear-Models/blob/master/module4-logistic-regression/LS_DS_214_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lambda School Data Science

*Unit 2, Sprint 1, Module 4*

---

# Logistic Regression


## Assignment 🌯

You'll use a [**dataset of 400+ burrito reviews**](https://srcole.github.io/100burritos/). How accurately can you predict whether a burrito is rated 'Great'?

> We have developed a 10-dimensional system for rating the burritos in San Diego. ... Generate models for what makes a burrito great and investigate correlations in its dimensions.

- [ ] Do train/validate/test split. Train on reviews from 2016 & earlier. Validate on 2017. Test on 2018 & later.
- [ ] Begin with baselines for classification.
- [ ] Use scikit-learn for logistic regression.
- [ ] Get your model's validation accuracy. (Multiple times if you try multiple iterations.)
- [ ] Get your model's test accuracy. (One time, at the end.)
- [ ] Commit your notebook to your fork of the GitHub repo.


## Stretch Goals

- [ ] Add your own stretch goal(s) !
- [ ] Make exploratory visualizations.
- [ ] Do one-hot encoding.
- [ ] Do [feature scaling](https://scikit-learn.org/stable/modules/preprocessing.html).
- [ ] Get and plot your coefficients.
- [ ] Try [scikit-learn pipelines](https://scikit-learn.org/stable/modules/compose.html).

In [1]:
%%capture
import sys

# If you're on Colab:
if 'google.colab' in sys.modules:
    DATA_PATH = 'https://raw.githubusercontent.com/LambdaSchool/DS-Unit-2-Linear-Models/master/data/'
    !pip install category_encoders==2.*

# If you're working locally:
else:
    DATA_PATH = '../data/'

In [2]:
# Load data downloaded from https://srcole.github.io/100burritos/
import pandas as pd
df = pd.read_csv(DATA_PATH+'burritos/burritos.csv')

In [3]:
# Derive binary classification target:
# We define a 'Great' burrito as having an
# overall rating of 4 or higher, on a 5 point scale.
# Drop unrated burritos.
df = df.dropna(subset=['overall'])
df['Great'] = df['overall'] >= 4

In [4]:
# Clean/combine the Burrito categories
df['Burrito'] = df['Burrito'].str.lower()

california = df['Burrito'].str.contains('california')
asada = df['Burrito'].str.contains('asada')
surf = df['Burrito'].str.contains('surf')
carnitas = df['Burrito'].str.contains('carnitas')

df.loc[california, 'Burrito'] = 'California'
df.loc[asada, 'Burrito'] = 'Asada'
df.loc[surf, 'Burrito'] = 'Surf & Turf'
df.loc[carnitas, 'Burrito'] = 'Carnitas'
df.loc[~california & ~asada & ~surf & ~carnitas, 'Burrito'] = 'Other'

In [5]:
# Drop some high cardinality categoricals
df = df.drop(columns=['Notes', 'Location', 'Reviewer', 'Address', 'URL', 'Neighborhood'])

In [6]:
# Drop some columns to prevent "leakage"
df = df.drop(columns=['Rec', 'overall'])

##  Split into Train, Validate, and Test Data
Train: before 2017

Validate: 2017

Test: after 2017

In [7]:
df.head()

Burrito       Date  Yelp  Google Chips  ...  Sushi  Avocado  Corn  Zucchini  Great
0  California  1/18/2016   3.5     4.2   NaN  ...    NaN      NaN   NaN       NaN  False
1  California  1/24/2016   3.5     3.3   NaN  ...    NaN      NaN   NaN       NaN  False
2    Carnitas  1/24/2016   NaN     NaN   NaN  ...    NaN      NaN   NaN       NaN  False
3       Asada  1/24/2016   NaN     NaN   NaN  ...    NaN      NaN   NaN       NaN  False
4  California  1/27/2016   4.0     3.8     x  ...    NaN      NaN   NaN       NaN   True

[5 rows x 59 columns]

In [8]:
df['Date'] = pd.to_datetime(df['Date'], infer_datetime_format=True)

In [9]:
train_cutoff = pd.to_datetime('2017-01-01')
val_cutoff = pd.to_datetime('2018-01-01')

train = df[df['Date'] < train_cutoff]
val = df[(df['Date'] >= train_cutoff) & (df['Date'] < val_cutoff)]
test = df[df['Date'] >= val_cutoff]

In [10]:
train.shape, val.shape, test.shape

((298, 59), (85, 59), (38, 59))

## Establish Baselines for Classification

In [11]:
target = 'Great'
y_train = train[target]
y_train.value_counts(normalize=True)

False    0.590604
True     0.409396
Name: Great, dtype: float64

In [12]:
majority_class = y_train.mode()
majority_class

0    False
dtype: bool

In [13]:
y_pred = [majority_class] * len(y_train)

In [14]:
from sklearn.metrics import accuracy_score

In [15]:
accuracy_score(y_train, y_pred )

0.5906040268456376

If we guessed that a burrito was rated as not being great, we would be correct in 59.1% of the cases.

## Logistic Regression

In [16]:
import category_encoders as ce
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LogisticRegressionCV
from sklearn.preprocessing import StandardScaler

/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [17]:
target = 'Great'
features = ['Burrito', 'Yelp', 'Google', 'Cost', 'Hunger', 'Tortilla', 'Temp', 'Meat', 'Fillings']

In [18]:
X_train = train[features]
y_train = train[target]
X_val = val[features]
y_val = val[target]

X_train.shape, y_train.shape, X_val.shape, y_val.shape

((298, 9), (298,), (85, 9), (85,))

In [19]:
# categorical encoding
encoder = ce.OneHotEncoder(use_cat_names=True)
X_train_encoded = encoder.fit_transform(X_train)
X_val_encoded = encoder.transform(X_val)

In [20]:
X_train_encoded.head()

Burrito_California  Burrito_Carnitas  Burrito_Asada  ...  Temp  Meat  Fillings
0                   1                 0              0  ...   5.0   3.0       3.5
1                   1                 0              0  ...   3.5   2.5       2.5
2                   0                 1              0  ...   2.0   2.5       3.0
3                   0                 0              1  ...   2.0   3.5       3.0
4                   1                 0              0  ...   5.0   4.0       3.5

[5 rows x 13 columns]

In [21]:
# fill in NaN values
imputer = SimpleImputer(strategy='mean')
X_train_imputed = imputer.fit_transform(X_train_encoded)
X_val_imputed = imputer.transform(X_val_encoded)

In [22]:
# scale
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train_imputed)
X_val_scaled = scaler.transform(X_val_imputed)

In [23]:
model = LogisticRegressionCV()
model.fit(X_train_scaled, y_train)
print('Validation Accuracy', model.score(X_val_scaled, y_val))

Validation Accuracy 0.8705882352941177


## Test Score for the Model

In [24]:
X_test = test[features]
y_test = test[target]

In [25]:
X_test_encoded = encoder.transform(X_test)
X_test_imputed = imputer.transform(X_test_encoded)
X_test_scaled = scaler.transform(X_test_imputed)

In [26]:
print('Validation Accuracy', model.score(X_test_scaled, y_test))

Validation Accuracy 0.7368421052631579
